File: python/Production/solver/RotationSequence.py
from production.utils.color import Color
from production.utils.face import Face
from production.utils.location import Location
from production.utils.rotation import Rotation
from production.utils.locationInFace import LocationInFace
from production.utils.direction import Direction
from production.utils.faceHandler import FaceHandler

class RotationSequence:

        def __init__(self, p_List=None):
            if p_List==None:
                self.c_array = list()
            else:
                self.c_array = p_List

        def print(self):
            for l_itr in self.c_array:
                l_itr.print()
            print("\n")

        def addRotation(self, p_rotation):
            self.c_array.append(p_rotation)

        def removeRotation(self):
            self.c_array.RemoveAt(len(self.c_array) - 1)

        def isRedundant(self, p_rotation):
            l_returnValue = False
            if len(self.c_array) > 0:
                l_lastFace = self.c_array[(len(self.c_array) - 1)].getFace()
                l_lastDirection = self.c_array[len(self.c_array) - 1].getDirection()
                # rotation is opposite to previous
                if self.c_array[len(self.c_array) - 1].getReverse().equals(p_rotation):
                    l_returnValue = True
                # previouswas opposite and previousgreater then current face
                if (p_rotation.getFace() == FaceHandler.getOpposite(l_lastFace) and (l_lastFace > p_rotation.getFace())):
                    l_returnValue = True
                # two clockwise rotation of same face
                if ((p_rotation.getFace() == l_lastFace) and (l_lastDirection == Direction.CW) and
                        p_rotation.getDirection() == Direction.CW):
                    l_returnValue = True
                #no three counter clockwise rotations
                if len(self.c_array) > 1:
                    if ((p_rotation.getFace() == l_lastFace) and (l_lastDirection == Direction.CCW) and
                            (p_rotation.getDirection() == Direction.CCW) and
                            (self.c_array[len(self.c_array) - 2].getFace() == l_lastFace) and (l_lastDirection == Direction.CCW) and
                            (self.c_array[len(self.c_array) - 2].getDirection() == Direction.CCW)):
                        l_returnValue = True
                else:
                    l_returnValue = False
            return l_returnValue

        def writeToFile(self, p_writer):
            for l_itr in self.c_array:
                l_itr.writeToFile(p_writer)
            p_writer.write("\n")
        
        def readFromFile(self, p_reader):
            l_rotation = Rotation()
            self.c_array.Clear()
            while l_rotation.readFromFile(p_reader):
                self.c_array.append((Rotation(l_rotation.getFace(), l_rotation.getDirection())))
            return (len(self.c_array) != 0)
       
        def getSubRotationLinkedList(self):
            return RotationSequence(self.c_array[1, len(self.c_array)])

        def size(self):
            return len(self.c_array)

        def  getFirstRotation(self):
            return self.c_array[0]
        
        def getRotation(self,p_index):
            return self.c_array[p_index]
        
        def isNotEmpty(self):
            return (len(self.c_array) > 0)

        def getCopy(self):
            l_rotationLinkedList = RotationSequence()
            for l_itr in self.c_array:
                l_rotationLinkedList.addRotation(l_itr)
            return l_rotationLinkedList

        def applyToRubik(self,p_rubik):
            for l_itr in self.c_array:
                p_rubik.rotateFace(l_itr.getFace(), l_itr.getDirection())

File: csharp/Production/solver/RotationSequence.cs
using CSharpRubikSolver;
using System;
using System.Collections.Generic;
using utils;
using cube;

namespace solver
{

    public class RotationSequence {

        private List<Rotation> c_array = new List<Rotation>();

        public RotationSequence() {
            c_array.Clear();
        }


        private RotationSequence(List<Rotation> p_List) {
            c_array = p_List;

        }

        public void print() {
         

            foreach (var l_itr in c_array)

                (l_itr as Rotation).print();
            Console.Write("\n");
        }

        public void addRotation(Rotation p_rotation) {
            c_array.Add(p_rotation);
        }

        public void removeRotation() {
            c_array.RemoveAt(c_array.Count - 1);
        }

        public Boolean isRedundant(Rotation p_rotation) {
            Boolean l_returnValue = false;

            Face l_lastFace;
            Direction l_lastDirection;
            if (c_array.Count > 0) {
                l_lastFace = c_array[(c_array.Count - 1)].getFace();
                l_lastDirection = c_array[c_array.Count - 1].getDirection();
                // new rotation is opposite to previous
                if (c_array[c_array.Count - 1].getReverse().equals(p_rotation))
                    l_returnValue = true;
                // previous face was opposite and previous face greater then current face
                if ((p_rotation.getFace() == FaceHandler.getOpposite(l_lastFace) && ((int)l_lastFace > (int)p_rotation.getFace())))
                    l_returnValue = true;
                // two clockwise rotation of same face
                if ((p_rotation.getFace() == l_lastFace) && (l_lastDirection == Direction.CW) &&
                        (p_rotation.getDirection() == Direction.CW))
                    l_returnValue = true;
                //no three counter clockwise rotations
                if (c_array.Count > 1) {
                    if ((p_rotation.getFace() == l_lastFace) && (l_lastDirection == Direction.CCW) &&
                            (p_rotation.getDirection() == Direction.CCW) &&
                            (c_array[c_array.Count - 2].getFace() == l_lastFace) && (l_lastDirection == Direction.CCW) &&
                            (c_array[c_array.Count - 2].getDirection() == Direction.CCW))
                        l_returnValue = true;
                }
            } else
                l_returnValue = false;
            return l_returnValue;

        }

        public void writeToFile(RubikFileWriter p_writer)
        {
            foreach (var l_itr in c_array)

                (l_itr as Rotation).writeToFile(p_writer);

            p_writer.write("\n");


        }
        public bool readFromFile(RubikFileReader p_reader)
        {
            Rotation l_rotation = new Rotation();

            c_array.Clear();
            while (l_rotation.readFromFile(p_reader))
                c_array.Add((new Rotation(l_rotation.getFace(), l_rotation.getDirection())));
            return !(c_array.Count == 0);
        }

       
        RotationSequence getSubRotationLinkedList() {
            return new RotationSequence(new List<Rotation>(c_array.GetRange(1, c_array.Count)));
        }

       public int size() {
            return c_array.Count;
        }

        Rotation getFirstRotation() {
            return c_array[0];
        }
        public Rotation getRotation(int p_index) {
            return c_array[p_index];
        }
        Boolean isNotEmpty() {

            return (c_array.Count > 0);
        }

        public RotationSequence getCopy() {
            RotationSequence l_rotationLinkedList = new RotationSequence();
           

        
            foreach (var l_itr in c_array)

                l_rotationLinkedList.addRotation(l_itr as Rotation);

            return l_rotationLinkedList;
        }

        public void applyToRubik(Cube p_rubik) {
           

              
            
            foreach (var l_itr in c_array)

                p_rubik.rotateFace(l_itr.getFace(), l_itr.getDirection());


        }
    }
}

